core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,79 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.syntax.group._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+class Perm private(val mapping: Map[Int, Int]) extends (Int => Int) {
+  assert(mapping.forall(Function.tupled(_ != _)), "Perm contains a redundant mapping.")
+  require(mapping.values.toSet == mapping.keySet, "Image and preimage must be the same.")
+
+  override def apply(k: Int): Int = mapping.getOrElse(k, k)
+
+  override def toString: String = {
+    mapping.toSeq.sorted
+      .map { case (k, v) => s"$k -> $v"}
+      .mkString("Perm(", ", ", ")")
+  }
+
+  def image: Set[Int] = mapping.keySet
+  def invert(k: Int): Int = mapping.find(_._2 == k).fold(k)(_._1)
+}
+
+object Perm {
+  def apply(mapping: Map[Int, Int]): Perm = {
+    new Perm(mapping.filter(Function.tupled(_ != _)))
+  }
+  def apply(pairs: (Int, Int)*): Perm = apply(Map(pairs: _*))
```

Comment:
filter out identity mappings here?

Reply:
If so, they'd be redundant with the other apply method.

Reply:
Oh, yes, I see